File: adventofcode.tests.csharp/2020/2020_13.cs
using System;
using System.Linq;
using Xunit;

namespace Year2020
{
    public static class AOC2020_13
    {
        public static Tuple<long, long> Solve(string[] lines)
        {
            long timestamp1 = long.Parse(lines[0]);
            var busInfos = lines[1].Split(',')
                .Select((id, idx) => Tuple.Create(id, idx))
                .Where(tup => tup.Item1 != "x")
                .Select(tup => new BusInfo(long.Parse(tup.Item1), tup.Item2))
                .ToArray();

            long minWait = long.MaxValue;
            long minId = long.MaxValue;
            foreach (int id in busInfos.Select(x => x.Id))
            {
                long mod = timestamp1 % id;
                long wait = mod == 0 ? 0 : id - mod;

                if (wait < minWait)
                {
                    minWait = wait;
                    minId = id;
                }
            }

            long ans1 = minId * minWait;

            long timestamp2 = 1;
            for (int i = 1; i <= busInfos.Length; i++)
            {
                // Increment amount = multiply together the IDs of the "locked in" buses. This is the gap we can jump each
                // time as we know everything inside this window is not a valid solution.
                long incAmt = busInfos.Take(i - 1).Aggregate(1L, (acc, b) => acc * b.Id);

                bool valid = false;
                while (!valid)
                {
                    valid = true;
                    for (int j = 0; j < i; j++)
                    {
                        // Check each potential timestamp to make sure it satisfies every bus under investigation.
                        long potentialTimestamp = timestamp2 + busInfos[j].Offset;
                        bool isMatch = potentialTimestamp % busInfos[j].Id == 0;
                        valid &= isMatch;
                    }
                    if (!valid)
                    {
                        // No match = skip the skippable gap and continue searching
                        timestamp2 += incAmt;
                    }
                }
            }

            long ans2 = timestamp2;

            return Tuple.Create(ans1, ans2);
        }

        record BusInfo(long Id, long Offset);
    }

    public class Day13
    {
        [Fact]
        public void Sample()
        {
            var lines = Utils.ReadInput("2020", "13", "sample");
            Utils.SolveAndValidate(Tuple.Create(295L, 1068781L), AOC2020_13.Solve, lines);
        }

        [Fact]
        public void Actual()
        {
            var lines = Utils.ReadInput("2020", "13", "actual");
            Utils.SolveAndValidate(Tuple.Create(104L, 842186186521918L), AOC2020_13.Solve, lines);
        }
    }
}

File: adventofcode.tests.fsharp/2020/2020_13.fs
namespace Year2020

module Day13 =
    open Utils
    open Xunit

    type BusInfo =
        { Id: int64
          Offset: int64 }

    let parseBusInfos (line : string) =
        line.Split(',')
        |> Array.mapi (fun offset id -> (offset, id))
        |> Array.filter (fun (_, id) -> id <> "x")
        |> Array.map (fun (offset, id) -> { Offset = int64 offset; Id = int64 id })

    let solve1 timestamp busInfos =
        let mutable minWait = System.Int64.MaxValue
        let mutable minId = System.Int64.MaxValue

        for info in busInfos do
            let modResult = timestamp % info.Id
            let wait = if modResult = 0L then 0L else info.Id - modResult

            if wait < minWait then
                minWait <- wait
                minId <- info.Id

        minId * minWait

    let solve2 busInfos =
        let len = Array.length busInfos
        let mutable timestamp = 1L

        for i in 1..len do
            // Increment amount = multiply together the IDs of the "locked in" buses. This is the gap we can jump
            // each time as we know everything inside this window is not a valid solution.
            let incAmt = (1L, busInfos |> Array.take (i-1)) ||> Array.fold (fun acc b -> acc * b.Id)

            let mutable valid = false
            while (not valid) do
                valid <- true;
                for j in 0..i-1 do
                    // Check each potential timestamp to make sure it satisfies every bus under investigation.
                    let potentialTimestamp = timestamp + busInfos.[j].Offset
                    let isMatch = potentialTimestamp % busInfos.[j].Id = 0L
                    valid <- valid && isMatch
                if not valid then
                    // No match = skip the skippable gap and continue searching
                    timestamp <- timestamp + incAmt
        timestamp

    let solve (lines : string[]) =
        let timestamp = lines.[0] |> int64
        let busInfos = lines.[1] |> parseBusInfos

        let ans1 = solve1 timestamp busInfos
        let ans2 = solve2 busInfos

        (ans1, ans2)

    [<Fact>]
    let Sample () =
        readInput "2020" "13" "sample" |> solveAndValidate (295L, 1068781L) solve

    [<Fact>]
    let Actual () =
        readInput "2020" "13" "actual" |> solveAndValidate (104L, 842186186521918L) solve